### PR TITLE
The Loadout Auto-Optimizer

### PR DESCRIPTION
**Pull request by robojumper**
Implements something like #9554. I got a bit carried away with complexity, there's a lot of duplication between LoadoutBuilder and auto-optimizer, the circular selector imports are getting worse and worse, this needs explanation, this needs a better workflow to actually find the sets discovered in the background in Loadout Optimizer, it probably invalidates the cache too often etc etc but it's a fun prototype.

### CODE REVIEW

**Comment by bhollis** [comment]
Seems to do something! What are the definitions of weak and strong upgrade?

**Comment by robojumper** [comment]
`AutoOptimizationResult` is commented with the definitions.

File: src/app/loadout-builder/auto-optimizer.ts [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+          mod.resolvedMod.plug.plugCategoryHash !== PlugCategoryHashes.EnhancementsV2General)
+    )
+    .map((mod) => mod.originalModHash);
+
+  if (!loadoutParameters?.statConstraints?.some((c) => c.minTier)) {
+    // If the user's loadout does not have any lower bounds, then we have no indication
+    // that the loadout parameters accurately reflect the stats the user cares about.
+    // Converting the existing set stats to tier minimums ensures that every weak upgrade is
+    // also a strong upgrade.
+    loadoutParameters.statConstraints = armorStats.map((statHash) => ({
```
Comment (robojumper) [inline src/app/loadout-builder/auto-optimizer.ts:261]: This should take existing stat order into account, at least.

File: src/app/loadout-builder/auto-optimizer.ts [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+  loadoutParameters.mods = originalLoadoutMods
+    .filter(
+      (mod) =>
+        // drop artifice mods (always picked automatically per set)
+        mod.resolvedMod.plug.plugCategoryHash !== PlugCategoryHashes.EnhancementsArtifice &&
+        // drop general mods if picked automatically
+        (!loadoutParameters?.autoStatMods ||
+          mod.resolvedMod.plug.plugCategoryHash !== PlugCategoryHashes.EnhancementsV2General)
+    )
+    .map((mod) => mod.originalModHash);
```
Comment (robojumper) [inline src/app/loadout-builder/auto-optimizer.ts:254]: Maybe we can get away with enabling auto mods if and only if the loadout has any stat mods, regardless the setting?

**Review by robojumper — COMMENTED** [review]
Some notes to self.

File: src/app/loadout-builder/auto-optimizer.ts [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+  const armorItems = items
+    .filter((i) => i.loadoutItem.equip && i.item.bucket.inArmor && i.item.energy)
+    .map((i) => i.item);
+  // A loadout must have 5 equipped armor 2.0 items for comparisons
+  // between LO runs and the current loadout to be meaningful
+  if (armorItems.length !== 5) {
+    return LoadoutError.NotAFullArmorSet;
+  }
+
```
Comment (robojumper) [inline src/app/loadout-builder/auto-optimizer.ts:203]: Maybe it should validate that the armor items match the search query too?

Comment (robojumper) [reply src/app/loadout-builder/auto-optimizer.ts:168]: Kind of tricky with the special behavior LO has for ignoring a search query in some cases...

**Comment by robojumper** [comment]
Superseded by #10012.

**Pull request closed without merging**